events/registry.go

```diff
@@ -0,0 +1,525 @@
+package events
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
+)
+
+type EventRegistry map[types.EventID]*EventType
+
+func (e EventRegistry) MarshalJSON() ([]byte, error) {
+	eventRegistryMap := make(map[string]*EventType)
+
+	for _, eventType := range e {
+		eventRegistryMap[eventType.Name] = eventType
+	}
+
+	return json.Marshal(eventRegistryMap)
+}
+
+func CreateEventRegistry(meta *types.Metadata) (EventRegistry, error) {
+	fieldStorage = make(map[int64]EventFieldType)
+
+	eventRegistry := make(map[types.EventID]*EventType)
+
+	for _, mod := range meta.AsMetadataV14.Pallets {
+		if !mod.HasEvents {
+			continue
+		}
+
+		eventsType, ok := meta.AsMetadataV14.EfficientLookup[mod.Events.Type.Int64()]
+
+		if !ok {
+			return nil, fmt.Errorf("events type %d not found for module '%s'", mod.Events.Type.Int64(), mod.Name)
+		}
+
+		if !eventsType.Def.IsVariant {
+			return nil, fmt.Errorf("events type %d for module '%s' is not a variant", mod.Events.Type.Int64(), mod.Name)
+		}
+
+		for _, eventVariant := range eventsType.Def.Variant.Variants {
+			eventID := types.EventID{byte(mod.Index), byte(eventVariant.Index)}
+
+			eventName := fmt.Sprintf("%s.%s", mod.Name, eventVariant.Name)
+
+			eventFields, err := getEventFields(meta, eventVariant.Fields)
+
+			if err != nil {
+				return nil, fmt.Errorf("couldn't get fields for event '%s': %w", eventName, err)
+			}
+
+			eventRegistry[eventID] = &EventType{
+				Name:   eventName,
+				Fields: eventFields,
+			}
+		}
+	}
+
+	return eventRegistry, nil
+}
+
+var (
+	fieldStorage fieldDecoderMap
+
+	fieldDecoderMtx sync.Mutex
+)
+
+type fieldDecoderMap map[int64]EventFieldType
+
+func (f fieldDecoderMap) getStoredEventFieldType(fieldName string, fieldType int64) (EventFieldType, bool) {
+	fieldDecoderMtx.Lock()
+	defer fieldDecoderMtx.Unlock()
+
+	if fieldDecoder, ok := f[fieldType]; ok {
+		if _, ok := fieldDecoder.(*RecursiveFieldType); ok {
+			fmt.Printf("Recursive type detected for field '%s'\n", fieldName)
+		}
+
+		return fieldDecoder, ok
+	}
+
+	// Ensure that a recursive type such as Xcm::TransferReserveAsset does not cause an infinite loop
+	// by adding the RecursiveFieldType the first time the field is encountered.
+	f[fieldType] = &RecursiveFieldType{}
```

Comment:
So since this registry stores types, then we only need to parse it once in case of recursion. And once we are parsing a complex/nested xcm message that will still work recursively, but the definition is always the same, right?

Reply:
In cases such as this one I think we should do another pass after we build the registry to find any recursive fields and replace them with the initial type.

Reply:
👍